Add two factor authentication flow

peppy opened the pull request:
Draft as this requires osu-web implementation.

Flows required:

- Endpoint to perform verification
- Endpoint to re-request verification code
- Response from a request (either error or otherwise) to let us know that verification hasn't been performed. ie. consider the case where the user does user/password login then quits the game before 2FA authenticating – to handle this, osu-web needs to let us know that we should show the flow again.

@nanaya is looking into the above requirements

- Closes https://github.com/ppy/osu/issues/20590
- [x] Depends on https://github.com/ppy/osu/pull/26724

---

https://github.com/ppy/osu/assets/191335/1012f75d-7c0b-484d-994a-c1beae88cc79

peppy committed:
Add two factor step to api state flow

peppy committed:
Setup basic form flow

peppy committed:
Add two factor auth form

peppy committed:
Adjust member ordering in `LoginForm`

peppy committed:
Add `APIAccess` flow for 2fa

peppy committed:
Update various components to handle new state

peppy committed:
Add test coverage of 2FA flow

peppy committed:
Tidy up localisations and connect missing links

bdach commented on osu.Game/Online/API/APIAccess.cs:
```diff
@@ -540,6 +574,11 @@ public enum APIState
+        /// <summary>
+        /// Waiting on second factor authentication.
+        /// </summary>
+        RequiresSecondFactorAuth,
```
Adding this to `APIState` is going to require a sweep of all `APIAccess` checks we have and seeing if they're broken by this or not.

A few places I could see being broken at a first glance:

https://github.com/ppy/osu/blob/87a76b9ee80055a85cb51a12e0a70083c1ba63e2/osu.Game/Online/OnlineViewContainer.cs#L63-L87
https://github.com/ppy/osu/blob/87a76b9ee80055a85cb51a12e0a70083c1ba63e2/osu.Game/Overlays/Comments/CommentEditor.cs#L181
https://github.com/ppy/osu/blob/87a76b9ee80055a85cb51a12e0a70083c1ba63e2/osu.Game/Overlays/AccountCreationOverlay.cs#L112-L128

and maybe things that use

https://github.com/ppy/osu/blob/87a76b9ee80055a85cb51a12e0a70083c1ba63e2/osu.Game/Online/API/APIAccess.cs#L461

peppy replied:
I think most of these are okay. Anything which was already valid in "connecting" should be valid in the 2FA state too. I'll fix the two switch based ones (which really should have throw-on-missing).

bdach commented on osu.Game/Overlays/Login/SecondFactorAuthForm.cs:
```diff
@@ -0,0 +1,117 @@
+                },
+            };
+
+            explainText.AddParagraph(UserVerificationStrings.BoxInfoCheckSpam);
+            // We can't support localisable strings with nested links yet. Not sure if we even can (probably need to allow markdown link formatting or something).
+            explainText.AddParagraph("If you can't access your email or have forgotten what you used, please follow the ");
+            explainText.AddLink(UserVerificationStrings.BoxInfoRecoverLink, $"{api.WebsiteRootUrl}/home/password-reset");
+            explainText.AddText(". You can also ");
```
Seems weird to intermingle these translatables with untranslatable and unarrangeable connecting strings. I'd just keep english for now.

peppy replied:
Yeah I guess. Kinda sad though.

bdach commented on osu.Game/Overlays/Login/SecondFactorAuthForm.cs:
```diff
@@ -0,0 +1,117 @@
+            {
+                if (code.NewValue.Length == 8)
+                {
+                    api.AuthenticateSecondFactor(code.NewValue);
```
This is presumably going to be a web request wherein the client sends the second factor code to web and validates it. If my thinking there is correct, this is gonna need a loading spinner for when the 2FA code is verified.

peppy replied:
It already changes the api state to "Connecting" again while performing the 2FA (you can test this, I added a sleep for testing purposes while in draft; you can also test success / failure as the "valid" code is all zeroes).

bdach commented on osu.Game/Online/API/APIAccess.cs:
```diff
@@ -208,11 +209,14 @@ private void attemptConnect()
                 try
                 {
                     authentication.AuthenticateWithLogin(ProvidedUsername, password);
+                    state.Value = APIState.RequiresSecondFactorAuth;
```
So this means _every_ user (who doesn't have their password enabled locally, I suppose) is going to have to 2FA verify on login? It's not like web wherein only certain user actions require verification?

peppy replied:
Yep, we're forcing it this time around at login. Seems more consistent.

bdach replied:
Yeah okay. I'm personally all for it because doing it as web does would be... rather painful...

That does beg the question though, what about users who have stored refresh tokens in config and as such will not go through the normal login flow? Is it presumed that the user is 2FA-verified throughout the entire lifetime of the refresh token? If so, are we going to have some possibility of revocation of the refresh token available to the user in event of account compromise?

peppy replied:
I've mentioned this in the OP:

> Response from a request (either error or otherwise) to let us know that verification hasn't been performed. ie. consider the case where the user does user/password login then quits the game before 2FA authenticating – to handle this, osu-web needs to let us know that we should show the flow again.

Basically there will be a case where web will let us know via a response to any API request that we need verification, which will cause the client to bring up the prompt.

This also means we could implement verify-when-required without any pain.

bdach reviewed (commented):
Aside from the test failures

peppy committed:
Update `APIState` checks

AAGaming00 commented:
Is proper TOTP also going to be considered in the future? I feel like it could be more convenient for a lot of people and is more secure.

peppy commented:
> Is proper TOTP also going to be considered in the future? I feel like it could be more convenient for a lot of people and is more secure.

https://github.com/ppy/osu-web/issues/5163 searching is not hard

bdach committed:
Merge branch 'master' into 2fa

bdach committed:
Fix login overlay test not clearing second auth factor

bdach committed:
Fix account creation overlay test not clearing second auth factor

bdach committed:
Fix `OsuGameTestScene` not clearing second auth factor

bdach committed:
Ensure all remaining test usages of `IAPIAccess.Login()` also authenticate with second factor

bdach committed:
Add request structures for verification endpoints

bdach committed:
Split `/me` request from `/users` requests

Them being together always bothered me and led to the abject failure
that is `APIUser` and its sprawl. Now that I'm about to add a flag that
is unique to `/me` for verification purposes, I'm not repeating the
errors of the past by adding yet another flag to `APIUser` that is never
present outside of a single usage context.

bdach committed:
Add `session_verified` attribute to `/me` response

bdach committed:
Implement verification from within client

bdach committed:
Hook up reissue request

bdach committed:
Split out raw websocket logic from conjoined notifications client contrivance

bdach committed:
Support verification via clicking link from e-mail

bdach committed:
Fix nullability inspection

bdach commented:
I've pushed some stuff here and the basic flows _appear_ to work. I'm not sure I'm super happy with the code, though.

First of all, the current "notification client" is almost *glued* to chat in its construction. Like have a look at `WebSocketNotificationClient` on master. It's half websocket handling logic, half chat specific stuff (things like fetching chat updates in `ConnectAsync()` via... calling base). I've attempted to basically step over that in https://github.com/ppy/osu/pull/25480/commits/62a0c236bc2eb706bb1f2840e80f7ed7df9c3b78 by splitting out the part I care about (the osu! ws glue code) from the rest but not sure how to feel about adding another layer to that lasagna.

Secondly there's https://github.com/ppy/osu/pull/25480/commits/e3eb7a8b4281c7d3b2170b10b2681cde8ee7f066: turns out that the previous flow of consumption of the notification websocket (a) was half-reliant on the user API state being online, (b) *required* to request the user's notifications to retrieve the *address of the notification websocket endpoint*:

https://github.com/ppy/osu/blob/b272d34960e701a844135cd558fab2fb95b5126d/osu.Game/Online/Notifications/WebSocket/WebSocketNotificationsClientConnector.cs#L30-L35

I'm not sure that's going to work if the user's session isn't verified, and it also seems turbo weird to do this. So I just hardcoded the websocket address. I'll likely want to ask the web team as to whether this is anywhere near acceptable.

Tests will probably fail. And there are no new tests exercising the new logic. I'll come back to that tomorrow.

nanaya commented:
the main idea for `/endpoint` is to avoid hardcoding the url in web (build/deployment constraints etc).

There's no particular reason why it requires authentication apart of there's no exclusion for it from requiring one and that it's not usable without authentication anyway.

bdach committed:
Merge branch 'decouple-notification-websocket-from-chat' into 2fa

bdach committed:
Handle forced logouts due to password change too

bdach committed:
Add basic testing of failure flow

bdach committed:
Merge branch 'decouple-notification-websocket-from-chat' into 2fa

bdach commented:
I think I'm *okay* with where this is at right now, but this is dependent on https://github.com/ppy/osu/pull/26724 now.

peppy committed:
Merge branch 'master' into 2fa

bdach committed:
Remove no longer relevant changes

bdach committed:
Fix `APIAccess` spamming requests while waiting for second factor

peppy committed:
Use `switch` instead of `if-else`

peppy committed:
Add loading layer when requesting a code reissue
